src/lightning/pytorch/cli.py

```diff
@@ -145,15 +145,26 @@ def add_lightning_class_args(
         ):
             if issubclass(lightning_class, Callback):
                 self.callback_keys.append(nested_key)
+
+            # NEW LOGIC: If subclass_mode=False and required=False, only add if config provides this key
+            if not subclass_mode and not required:
+                config_path = f"{self.subcommand}.{nested_key}" if getattr(self, "subcommand", None) else nested_key
+                config = getattr(self, "config", {})
+                if not any(k.startswith(config_path) for k in config):
+                    # Skip adding class arguments
+                    return []
```

Comment:
This method is only called on parser construction, not during parsing. So at this point there isn't any config. Because of the `getattr(self, "config", {})` this doesn't fail but it will never do anything.